### PR TITLE
Multi-fidelity Bayesian optimization notebook

### PR DESCRIPTION
**Pull request by marpulli**
Multi-fidelity Bayesian optimization notebook + multi information source entropy search acquisition function

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 1]
```diff
@@ -96,12 +96,14 @@ def Y(self) -> np.ndarray:
         return self.model.Y
 
 
-class GPyMultiOutputWrapper(IModel, IDifferentiable, ICalculateVarianceReduction):
+class GPyMultiOutputWrapper(IModel, IDifferentiable, ICalculateVarianceReduction, IEntropySearchModel):
```
Comment (javiergonzalezh) [inline emukit/model_wrappers/gpy_model_wrappers.py]: What if I want to use a multiple-output GP with an acquisition that is not ES. Would this require to always have the ES interface?

Comment (marpulli) [reply emukit/model_wrappers/gpy_model_wrappers.py]: I don't fully understand the comment but this change allows you to use entropy search but does not limit you doing anything else.

**Comment by mmahsereci** [comment]
Could you include the BO high-fidelity-only baseline in the notebook? I find it less convincing without it since it does not showcase why one would want to do MF instead of vanilla BO.

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
...x/TpE6hduyEtWnTwdVhZQgm95Gpt2rShTZs2vg7jmlOwYEFUbiciIrmVtbB5MxQoACdPHqBAgQD69x+dK1vnQQm9iIiIiOQyBw86484HB0OxYpX49NN1vg4pS2mUGxERERHJNaKjYeNGKFEC9u/fydmzp8iXL1+ufnBi7j2yLKIyBZGrp78fERHJalFRzru/P7z00oP06RNKfHy8b4PKYkroM8Df3z/VJ26KyJVduHABf//c9TAPERG5dhw7Bnv3Oq3zf/+9iZ9++pYOHfrm6tZ5UEKfIcHBwezbt4/z58+rpVEkA6y1nD9/nn379hEcHOzrcEREJBeKjYX166FYMTAGPv74VQoWDKR792G+Di3LqVNsBhQtWhSA/fv3ExMT4+NoRHIWf39/ypYtm/B3JCIi4k07dsClS1C0KBw+vI/vvvuEbt2GULx4KV+HluWU0GdQ0aJFlZCIiIiIXENOn4Zt26B0aefzjz8uID4+jvvuG+nbwLKJEnoRERERybHi42HTJggMBHepfLdug2nWrB0VKoT4NLbsohp6EREREcmx9u6FkychKMj5HBvrlEXnlWQelNCLiIiISA514QJs2QIlSzqf4+LiuOeeG5g+fYJvA8tmSuhFREREJMex1knm8+d3XgArV37Drl1RVKlSy7fBZTMl9CIiIiKS4xw+DAcOQPHil6fNnv1fypatTFhYV98F5gNK6EVEREQkR7l0CTZudB4g5bZt2wZWr15Cjx4Pkj9/3hr3RQm9iIiIiOQo27ZBXBwEBFyeNnv2mwQEFKJLl4G+C8xH8tbli4iIiIjkaCdPws6dkPzB4w888G+aNm2XJx4klZwSehERERHJEeLiYMMGKFIEjEk6r0KFkDw1VGViKrkRERERkRxh1y44e9Z5iJRbbGwsY8f2ZePGX3wXmI/liITeGFPJGPOmMSbSGHPeGGONMSEelitojHnFGHPAGHPBtXzL7I9YRERERLzp7FmIioJSySpqli2bx4IFH3H06EHfBHYNyBEJPVAT6AmcAFamsdz7wCBgDNAROAB8b4xplOURioiIiEiWsBY2bYKCBcHPL+m8WbP+S8WK1bjtto6+Ce4akFMS+hXW2rLW2juBzz0tYIxpCNwHjLTWvmut/R/ORcBu4PnsC1VEREREvGn/fjh+HIoWTTp969a1rFv3Iz16PIRf8kw/D8kRCb21
```
Comment (javiergonzalezh) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Let's also cite this one here
https://papers.nips.cc/paper/5086-multi-task-bayesian-optimization.pdf

File: emukit/model_wrappers/gpy_model_wrappers.py (hunk 1 above)
Comment (javiergonzalezh) [reply emukit/model_wrappers/gpy_model_wrappers.py]: OK, fine, it is GPy specific so no issues.

File: emukit/bayesian_optimization/acquisitions/entropy_search.py [hunk 3]
```diff
@@ -208,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
...
+        :param space: Parameter space of the input domain
+        :param target_information_source_index: The index of the information source we want to minimise
+        :param num_samples: Integer determining how many samples to draw for each candidate input
+        :param num_representer_points: Integer determining how many representer points to sample
+        :param burn_in_steps: Integer that defines the number of burn-in steps when sampling the representer points
+        """
+
+        # Find information source parameter in parameter space
+        info_source_parameter = None
+        self.fidelity_idx = None
```
Comment (javiergonzalezh) [inline emukit/bayesian_optimization/acquisitions/entropy_search.py]: Is it right to call this fidelity if this class is general for muti output/fidelity cases?

File: emukit/bayesian_optimization/acquisitions/entropy_search.py [hunk 4]
```diff
@@ -208,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
...
+        for i, param in enumerate(space.parameters):
+            if isinstance(param, InformationSourceParameter):
+                info_source_parameter = param
+                self.fidelity_idx = i
```
Comment (javiergonzalezh) [inline emukit/bayesian_optimization/acquisitions/entropy_search.py]: same here, shall we use fidelity?

**Comment by marpulli** [comment]
@mmahsereci I'll add a high fidelity only baseline at the end of the notebook

**Comment by marpulli** [comment]
- Added a comparison to high fidelity only at the end of the notebook.
- Changed fidelity -> index in entropy search code
- Added Multi-Task Bayes opt reference to notebook

File: emukit/bayesian_optimization/acquisitions/entropy_search.py [hunk 5]
```diff
@@ -214,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
...
         """Returns that this acquisition has gradients"""
         return False
+
+
+class MultiInformationSourceEntropySearch(EntropySearch):
+    """
+    Entropy search acquisition for multi-information source problems where the objective function is the output of one
+    of the information sources. The other information sources provide auxiliary information about the objective function
+    """
+    def __init__(self, model: Union[IModel, IEntropySearchModel], space: ParameterSpace,
```
Comment (apaleyes) [inline emukit/bayesian_optimization/acquisitions/entropy_search.py]: Any chance you could move part of this __init__ in a separate method, to shorten it a bit? Maybe identifying info source param piece? Maybe creation of proposal function?

File: emukit/bayesian_optimization/acquisitions/entropy_search.py [hunk 6]
```diff
@@ -214,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
     def has_gradients(self) -> bool:
         """Returns that this acquisition has gradients"""
         return False
+
+
+class MultiInformationSourceEntropySearch(EntropySearch):
```
Comment (apaleyes) [inline emukit/bayesian_optimization/acquisitions/entropy_search.py]: The fact that we can reuse stuff like that is just great!

File: notebooks/Emukit-multifidelity-tutorial.ipynb [hunk 7]
```diff
@@ -847,6 +847,13 @@
     "\n",
     "- Perdikaris, P., Raissi, M., Damianou, A., Lawrence, N.D. and Karniadakis, G.E., 2017. *Nonlinear information fusion algorithms for data-efficient multi-fidelity modelling.* Proc. R. Soc. A, 473(2198), p.20160751."
    ]
+  },
+  {
```
Comment (apaleyes) [inline notebooks/Emukit-multifidelity-tutorial.ipynb]: Empty cell?

File: tests/emukit/bayesian_optimization/test_entropy_search.py [hunk 8]
```diff
@@ -55,3 +66,12 @@ def test_innovations(acquisition):
     dm, dv = acquisition._innovations(x)
     assert np.any(np.abs(dm) > 1e-3)
     assert np.any(np.abs(dv) > 1e-3)
+
+
+def test_multi_information_source_entropy_search_shape(multi_source_acquisition):
+    rng = np.random.RandomState(43)
+
+    x_test = np.concatenate([rng.rand(2, 1), np.array([[0], [1]])], axis=1)
```
Comment (apaleyes) [inline tests/emukit/bayesian_optimization/test_entropy_search.py]: Fixing random seed and calling random is just as good as hardcoding numbers. The latter however is simpler.

That is, unless random is also called somewhere down the line, but does not look like there is any

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+    "\n",
+    "The low fidelity approximation of this function is given by:\n",
+    "\n",
+    "$$\n",
+    "f_{low}(x) = 0.5 f_{high}(x) + 10 (x - 0.5) + 5\n",
+    "$$\n",
+    "\n",
+    "Both are defined in the interval [0, 1].\n",
+    "\n",
+    "The high fidelity function is assigned a cost of $3$, whereas the low fidelity function is assigned a cost of $1$.\n",
```
Comment (apaleyes) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: "is assigned a cost of X" is a bit unclear. "Each single evaluation of f costs X" would be clearer to me

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+    "y_high = forrester_fcn_high(x_high)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define Model\n",
+    "\n",
+    "We will use the linear multi-fidelity model defiend in Emukit. In this model, the high-fidelity function is modelled as a scaled sum of the low-fidelity function plus an error term:\n",
```
Comment (apaleyes) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]:
```suggestion
    "We will use the linear multi-fidelity model defined in Emukit. In this model, the high-fidelity function is modelled as a scaled sum of the low-fidelity function plus an error term:\n",
```

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
...ZtypQpA0CtWrWIi4vju+++o3v37tmWI4mIyIXD44EVK2DnzuDsAAtumMq3336EsmUr0KvXE/4OJ8/0bF3yZOjQodx777289dZbdO3alQ0bNjBu3Lgs23Xt2pXZs2ezf/9++vbtS1xcHE8++SR79+7lL3/5y3mL78MPP6Rnz568/PLLdOvWjfXr1/POO+8AUKpUqfN2Xl8iIiL4/vvvueSSS+jfvz99+vTh0ksvzTCSTao333yTd999l++//55bbrmFm266iREjRhAdHU3NmjVzfM6iRYvywQcfsHPnTq6//nquuuoqpk+fnmGb1E656gwrIiKpk0alDk0ZrLZu/ZM//ljGgAEvULx4CX+Hk2fGWuvvGAqN2NhY66vGOdWqVavSJuqRwPXSSy8xdOhQtm/frhZp4Oabb2bFihVpE4Kdb/o7EhEpnKyF33+HTZuCb5x5Xw4c2EOJEmUoUqTgC1b27YO4uPDl1p5omB/HU8mNBLUpU6awbt06GjZ0fy9z587lzTff5K677rqgk3lrLb/88gsLFixgypQpabPsiojIhclaWL36wkjmt23bQJUqMZQpEzydA5TQS1ArUaIEn3/+OS+99BLHjh2jevXqPProozz33HP+Ds2vkpKSaNasGSVKlGDAgAH069fP3yGJiIgfrVsH69dDxYrBncwfP36Ue+5pyvXX38ITT4z0dzj5Rgm9BLW2bdvStm1bf4dR6ISHh6NyOxERAdi4EdauDf6WeYBJk95l//7ddOiQt1HkCht1ihURERG5QG3d6jrBRkdDsE9DcvjwQf7zn9e49tpONGqUs9HiAkWQf3UiIiIi4svOnW6s+fLlITTU39GcfxMmvEli4gEGDXrJ36HkOyX0uaQyBZFzp78fEZHCYd8+WLwYypYFPwzyUuA8Hg9z5nzF9dffQoMGV/g7nHx3AXyF+ado0aIcP36cyMhIf4ciEpCOHz9O0aJF/R2GiMgF7dAhWLAASpWCC+U/ySEhIYwfv5AjRw75O5TzQi30uVChQgW2bdvGsWPH1NIokgvWWo4dO8a2bduoEMwzlYiIFHJHjsCvv0JUFISF+TuagnH48EGSkk5QtGixoBqqMj210OdCyZIlAdi+fTunTp3yczQigaVo0aJUrFgx7e9IREQK1vHjrmW+WDGIiPB3NAVn1Kjn+PHHb/jss98JCwv3dzjnhRL6XC
```
Comment (apaleyes) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: A sentence about why and how we subscribe to an event to make that happen would be helpful

Comment (apaleyes) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Very nice to see events in action

Comment (apaleyes) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: This sentence lacks a verb somehow

Comment (apaleyes) [inline notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: A final sentence analyzing the comparison figure would be nice.

File: emukit/bayesian_optimization/acquisitions/entropy_search.py (hunk 5 above)
Comment (marpulli) [reply emukit/bayesian_optimization/acquisitions/entropy_search.py]: Done

File: notebooks/Emukit-multifidelity-tutorial.ipynb (hunk 7 above)
Comment (marpulli) [reply notebooks/Emukit-multifidelity-tutorial.ipynb]: Removed this file from PR

File: tests/emukit/bayesian_optimization/test_entropy_search.py (hunk 8 above)
Comment (marpulli) [reply tests/emukit/bayesian_optimization/test_entropy_search.py]: Done

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb (hunk 11 above)
Comment (marpulli) [reply notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Fixed

Comment (marpulli) [reply notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Done

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb (hunk 9 above)
Comment (marpulli) [reply notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Changed

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb (hunk 11 above)
Comment (marpulli) [reply notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Added

File: emukit/bayesian_optimization/acquisitions/entropy_search.py [hunk 12]
```diff
@@ -214,3 +215,85 @@ def _innovations(self, x: np.ndarray) -> tuple:
...
+        :param model: Gaussian process model of the objective function that implements IEntropySearchModel
+        :param space: Parameter space of the input domain
+        :param target_information_source_index: The index of the information source we want to minimise
+        :param num_samples: Integer determining how many samples to draw for each candidate input
+        :param num_representer_points: Integer determining how many representer points to sample
+        :param burn_in_steps: Integer that defines the number of burn-in steps when sampling the representer points
+        """
+
+        # Find information source parameter in parameter space
+        info_source_parameter, source_idx = MultiInformationSourceEntropySearch._find_source_parameter(space)
```
Comment (apaleyes) [inline emukit/bayesian_optimization/acquisitions/entropy_search.py]: Underscore normally means private member, so calling it like that would be highly not advisable. Make "find_source_parameter" a package level function and use it like that

File: notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb (hunk 9 above)
Comment (apaleyes) [reply notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Is it?

Comment (marpulli) [reply notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb]: Weird, I'm sure I changed them in the github editor. I'll try again

**Pull request merged**